codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoderForBuffer.java

```diff
@@ -0,0 +1,683 @@
+            channelInputClosed(context, false);
+        }
+    }
+
+    private void channelInputClosed(ByteToMessageDecoderContext ctx, boolean callChannelInactive) {
+        try {
+            channelInputClosed(ctx);
+        } catch (DecoderException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new DecoderException(e);
+        } finally {
+            if (cumulation != null) {
+                cumulation.close();
+                cumulation = null;
+            }
+            if (ctx.fireChannelReadCallCount() > 0) {
+                ctx.reset();
+                // Something was read, call fireChannelReadComplete()
+                ctx.fireChannelReadComplete();
+            }
+            if (callChannelInactive) {
+                ctx.fireChannelInactive();
+            }
+        }
+    }
+
+    /**
+     * Called when the input of the channel was closed which may be because it changed to inactive or because of
+     * {@link ChannelInputShutdownEvent}.
+     */
+    void channelInputClosed(ByteToMessageDecoderContext ctx) throws Exception {
+        if (cumulation != null) {
+            callDecode(ctx, cumulation);
+            // If callDecode(...) removed the handle from the pipeline we should not call decodeLast(...) as this would
+            // be unexpected.
+            if (!ctx.isRemoved()) {
+                // Use Unpooled.EMPTY_BUFFER if cumulation become null after calling callDecode(...).
+                // See https://github.com/netty/netty/issues/10802.
+                Buffer buffer = cumulation == null ? ctx.bufferAllocator().allocate(0) : cumulation;
+                decodeLast(ctx, buffer);
+            }
+        } else {
+            decodeLast(ctx, ctx.bufferAllocator().allocate(0));
+        }
+    }
+
+    /**
+     * Called once data should be decoded from the given {@link Buffer}. This method will call
+     * {@link #decode(ChannelHandlerContext, Buffer)} as long as decoding should take place.
+     *
+     * @param ctx the {@link ChannelHandlerContext} which this {@link ByteToMessageDecoderForBuffer} belongs to
+     * @param in  the {@link Buffer} from which to read data
+     */
+    void callDecode(ByteToMessageDecoderContext ctx, Buffer in) {
+        try {
+            while (in.readableBytes() > 0 && !ctx.isRemoved()) {
+
+                int oldInputLength = in.readableBytes();
+                int numReadCalled = ctx.fireChannelReadCallCount();
+                decodeRemovalReentryProtection(ctx, in);
+
+                // Check if this handler was removed before continuing the loop.
+                // If it was removed, it is not safe to continue to operate on the buffer.
+                //
+                // See https://github.com/netty/netty/issues/1664
+                if (ctx.isRemoved()) {
+                    break;
+                }
+
+                if (numReadCalled == ctx.fireChannelReadCallCount()) {
```

Comment:
I know this is just copied from the original B2MD, but… why is it counting read calls and not write calls? And why is movement of the buffer read-offset not enough?